xarray/core/variable.py

```diff
@@ -2418,6 +2418,10 @@ def _inplace_binary_op(self, other, f):
         if dims != self.dims:
             raise ValueError("dimensions cannot change for in-place operations")
         with np.errstate(all="ignore"):
+            # Give a chance to Variable.values.setter to throw error if needed
+            # without updating self_data inplace
+            if isinstance(self_data, np.ndarray):
+                self.values = f(self_data.copy(), other_data)
```

Comment:
Defensively copying xarray.Variable data would negate the main purpose of in-place binary arithmetic, which is to avoid copies.

For fixing https://github.com/pydata/xarray/issues/5036, what about instead overriding `_inplace_binary_op` on the `IndexVariable` subclass? I think we could just make it raise a `TypeError` instead of attempting to do the computation.